File: modules/wem-api/src/main/java/com/enonic/wem/api/content/page/PageRegions.java
package com.enonic.wem.api.content.page;


import com.enonic.wem.api.content.page.region.AbstractRegions;

public final class PageRegions
    extends AbstractRegions
{
    private PageRegions( final Builder builder )
    {
        super( builder );
    }

    public static Builder newPageRegions()
    {
        return new Builder();
    }

    public static Builder newPageRegions( final PageRegions source )
    {
        return new Builder( source );
    }

    public PageRegions copy()
    {
        return PageRegions.newPageRegions( this ).build();
    }

    public static class Builder
        extends AbstractRegions.Builder<Builder>
    {
        private Builder( final PageRegions source )
        {
            super( source );
        }

        private Builder()
        {
            // Default
        }

        public PageRegions build()
        {
            return new PageRegions( this );
        }
    }
}

File: modules/wem-webapp/src/main/webapp/admin/common/js/content/page/PageRegions.ts
module api.content.page {

    export class PageRegions extends AbstractRegions implements api.Equitable {

        constructor(builder: PageRegionsBuilder) {
            super(builder.regions);
        }

        equals(o: api.Equitable): boolean {

            if (!(o instanceof PageRegions)) {
                return false;
            }

            return super.equals(o);
        }
    }

    export class PageRegionsBuilder {

        regions: region.Region[] = [];

        fromJson(regionsJson: api.content.page.region.RegionJson[]): PageRegionsBuilder {

            regionsJson.forEach((regionJson: api.content.page.region.RegionJson) => {

                var regionPath = new RegionPath(null, regionJson.name);

                var regionBuilder = new region.RegionBuilder().
                    setName(regionJson.name).
                    setPath(regionPath);

                regionJson.components.forEach((componentJson: api.content.page.PageComponentTypeWrapperJson) => {
                    var pageComponent = PageComponentFactory.createFromJson(componentJson, regionPath);
                    regionBuilder.addComponent(pageComponent);
                });


                this.addRegion(regionBuilder.build());
            });
            return this;
        }

        addRegion(value: region.Region): PageRegionsBuilder {
            this.regions.push(value);
            return this;
        }

        public build(): PageRegions {
            return new PageRegions(this);
        }
    }
}